File: app/lib/api/protocol.dart
// Wire protocol between the Flutter app and the Claude Companion server.

sealed class OutgoingMessage {
  Map<String, dynamic> toJson();
}

class InitMessage extends OutgoingMessage {
  final String cwd;
  final String permissionMode;
  InitMessage({required this.cwd, this.permissionMode = 'acceptEdits'});

  @override
  Map<String, dynamic> toJson() => {
        'type': 'init',
        'cwd': cwd,
        'permission_mode': permissionMode,
      };
}

class UserTextMessage extends OutgoingMessage {
  final String text;
  UserTextMessage(this.text);

  @override
  Map<String, dynamic> toJson() => {'type': 'user_message', 'text': text};
}

class InterruptMessage extends OutgoingMessage {
  @override
  Map<String, dynamic> toJson() => {'type': 'interrupt'};
}

class PingMessage extends OutgoingMessage {
  @override
  Map<String, dynamic> toJson() => {'type': 'ping'};
}

abstract class IncomingMessage {
  static IncomingMessage fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String?;
    switch (type) {
      case 'session_ready':
        return SessionReady(
          sessionKey: json['session_key'] as String? ?? '',
          cwd: json['cwd'] as String? ?? '',
          permissionMode: json['permission_mode'] as String? ?? 'acceptEdits',
          busy: json['busy'] as bool? ?? false,
        );
      case 'assistant':
        return AssistantMsg(
          model: json['model'] as String?,
          content: ((json['content'] as List?) ?? [])
              .map((b) => ContentBlock.fromJson(Map<String, dynamic>.from(b)))
              .toList(),
          timestamp: (json['timestamp'] as num?)?.toInt(),
        );
      case 'user':
        return UserMsg(
          content: ((json['content'] as List?) ?? [])
              .map((b) => ContentBlock.fromJson(Map<String, dynamic>.from(b)))
              .toList(),
          timestamp: (json['timestamp'] as num?)?.toInt(),
        );
      case 'result':
        return ResultMsg(
          durationMs: (json['duration_ms'] as num?)?.toInt(),
          totalCostUsd: (json['total_cost_usd'] as num?)?.toDouble(),
          sessionId: json['session_id'] as String?,
          numTurns: (json['num_turns'] as num?)?.toInt(),
          isError: (json['is_error'] as bool?) ?? false,
          timestamp: (json['timestamp'] as num?)?.toInt(),
        );
      case 'system':
        return SystemMsg(
          subtype: json['subtype'] as String?,
          data: Map<String, dynamic>.from(json['data'] ?? {}),
        );
      case 'error':
        return ErrorMsg(message: json['message'] as String? ?? 'Unknown error');
      case 'pong':
        return PongMsg();
      case 'stream_block_start':
        return StreamBlockStart(
          index: (json['index'] as num?)?.toInt() ?? 0,
          kind: json['kind'] as String? ?? 'unknown',
        );
      case 'stream_delta':
        return StreamDelta(
          index: (json['index'] as num?)?.toInt() ?? 0,
          kind: json['kind'] as String? ?? 'text',
          text: json['text'] as String? ?? '',
        );
      case 'stream_block_stop':
        return StreamBlockStop(index: (json['index'] as num?)?.toInt() ?? 0);
      case 'compact_boundary':
        return CompactBoundaryMsg(
          trigger: json['trigger'] as String?,
          preTokens: (json['pre_tokens'] as num?)?.toInt(),
          postTokens: (json['post_tokens'] as num?)?.toInt(),
          durationMs: (json['duration_ms'] as num?)?.toInt(),
          timestamp: (json['timestamp'] as num?)?.toInt(),
        );
      default:
        return UnknownMsg(raw: json);
    }
  }
}

/// jsonl 里的 `system / compact_boundary` 行：会话上下文被压缩的边界标记。
/// 之前的消息在 jsonl 中仍在，但 SDK 在 resume 时会跳过，所以视觉上"消息消失"。
/// 我们在历史流里画一条分隔线，告诉用户这里发生了什么。
class CompactBoundaryMsg extends IncomingMessage {
  final String? trigger; // 'manual' | 'auto' | null
  final int? preTokens;
  final int? postTokens;
  final int? durationMs;
  final int? timestamp;
  CompactBoundaryMsg({
    this.trigger,
    this.preTokens,
    this.postTokens,
    this.durationMs,
    this.timestamp,
  });
}

class StreamBlockStart extends IncomingMessage {
  final int index;
  final String kind;
  StreamBlockStart({required this.index, required this.kind});
}

class StreamDelta extends IncomingMessage {
  final int index;
  final String kind;
  final String text;
  StreamDelta({required this.index, required this.kind, required this.text});
}

class StreamBlockStop extends IncomingMessage {
  final int index;
  StreamBlockStop({required this.index});
}

class SessionReady extends IncomingMessage {
  final String sessionKey;
  final String cwd;
  final String permissionMode;
  /// 服务端 attach 已有 session 时为 true：当前有 in-flight 的流式响应，
  /// 客户端应立即恢复 spinner / streaming UI，不必等下一个事件。
  final bool busy;
  SessionReady({
    required this.sessionKey,
    required this.cwd,
    required this.permissionMode,
    this.busy = false,
  });
}

class AssistantMsg extends IncomingMessage {
  final List<ContentBlock> content;
  final String? model;
  final int? timestamp; // epoch ms; null when server didn't tag it
  AssistantMsg({required this.content, this.model, this.timestamp});
}

class UserMsg extends IncomingMessage {
  final List<ContentBlock> content;
  final int? timestamp;
  UserMsg({required this.content, this.timestamp});
}

class ResultMsg extends IncomingMessage {
  final int? durationMs;
  final double? totalCostUsd;
  final String? sessionId;
  final int? numTurns;
  final bool isError;
  final int? timestamp;
  ResultMsg({
    this.durationMs,
    this.totalCostUsd,
    this.sessionId,
    this.numTurns,
    this.isError = false,
    this.timestamp,
  });
}

class SystemMsg extends IncomingMessage {
  final String? subtype;
  final Map<String, dynamic> data;
  SystemMsg({this.subtype, required this.data});
}

class ErrorMsg extends IncomingMessage {
  final String message;
  ErrorMsg({required this.message});
}

class PongMsg extends IncomingMessage {}

class UnknownMsg extends IncomingMessage {
  final Map<String, dynamic> raw;
  UnknownMsg({required this.raw});
}

sealed class ContentBlock {
  static ContentBlock fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String?;
    switch (type) {
      case 'text':
        return TextBlock(text: json['text'] as String? ?? '');
      case 'thinking':
        return ThinkingBlock(text: json['text'] as String? ?? '');
      case 'tool_use':
        return ToolUseBlock(
          id: json['id'] as String? ?? '',
          name: json['name'] as String? ?? '',
          input: Map<String, dynamic>.from(json['input'] ?? {}),
        );
      case 'tool_result':
        return ToolResultBlock(
          toolUseId: json['tool_use_id'] as String? ?? '',
          content: json['content'],
          isError: (json['is_error'] as bool?) ?? false,
        );
      default:
        return UnknownBlock(raw: json);
    }
  }
}

class TextBlock extends ContentBlock {
  final String text;
  TextBlock({required this.text});
}

class ThinkingBlock extends ContentBlock {
  final String text;
  ThinkingBlock({required this.text});
}

class ToolUseBlock extends ContentBlock {
  final String id;
  final String name;
  final Map<String, dynamic> input;
  ToolUseBlock({required this.id, required this.name, required this.input});
}

class ToolResultBlock extends ContentBlock {
  final String toolUseId;
  final dynamic content;
  final bool isError;
  ToolResultBlock({required this.toolUseId, required this.content, required this.isError});
}

class UnknownBlock extends ContentBlock {
  final Map<String, dynamic> raw;
  UnknownBlock({required this.raw});
}

File: packages/shared/src/protocol.ts
/**
 * Wire protocol between server and clients (Flutter app, Web admin).
 * Stable contract — any change needs migration in both client codebases.
 */

// ============== Common ==============

export type PermissionMode = 'default' | 'acceptEdits' | 'plan' | 'bypassPermissions';

// ============== Health ==============

export interface HealthResponse {
  status: string;
  version: string;
  hostname: string;
  serverId?: string;
  pairingOpen?: boolean;
}

// ============== Pairing ==============

// POST /admin/pair-window — requires adminToken
export interface PairWindowRequest {}
export interface PairWindowResponse { pin: string; expiresAt: number }

// POST /pair/start — no auth; PIN is the out-of-band credential
export interface PairStartRequest { deviceId: string; deviceName: string; pin: string }
export type PairStartResponse =
  | { ok: true; deviceToken: string; serverId: string }
  | { ok: false; error: 'bad_pin' | 'pairing_closed' | 'rate_limited' };

// POST /pair/qr-claim — requires adminToken
export interface PairQrClaimRequest { deviceId: string; deviceName: string }
export interface PairQrClaimResponse { deviceToken: string; serverId: string }

// GET /admin/devices — list; DELETE /admin/devices/:id — revoke; requires adminToken
export interface PairedDevice {
  deviceId: string;
  name: string;
  pairedAt: number;  // epoch ms
  lastSeen: number | null;
}

// ============== Chat WebSocket: /ws/session (web admin only) ==============
// The Flutter app has migrated to REST + SSE. This union type is kept only for
// the web admin's wsChat.ts until that client is also migrated.

export type ChatClientMessage =
  | { type: 'init'; cwd: string; permission_mode?: PermissionMode; resume?: string; model?: string }
  | { type: 'user_message'; text: string }
  | { type: 'set_model'; model: string }
  | { type: 'set_permission_mode'; mode: PermissionMode }
  | { type: 'interrupt' }
  | { type: 'ping' };

/** Available Claude models the client can pick. Keep in sync with App + Web. */
export const KNOWN_MODELS = [
  { id: 'claude-sonnet-4-6', label: 'Sonnet 4.6', tier: 'fast' },
  { id: 'claude-opus-4-7', label: 'Opus 4.7', tier: 'powerful' },
  { id: 'claude-haiku-4-5', label: 'Haiku 4.5', tier: 'cheap' },
] as const;

export type ChatServerMessage =
  | { type: 'session_ready'; session_key: string; cwd: string; permission_mode: PermissionMode; resumed?: string | null; busy?: boolean }
  | { type: 'assistant'; model?: string; content: ContentBlock[]; timestamp?: number }
  | { type: 'user'; content: ContentBlock[]; timestamp?: number }
  | { type: 'system'; subtype?: string; data?: unknown; timestamp?: number }
  | { type: 'result'; subtype?: string; duration_ms?: number; duration_api_ms?: number; is_error: boolean; num_turns?: number; session_id?: string; total_cost_usd?: number; usage?: unknown; timestamp?: number }
  | { type: 'stream_block_start'; index: number; kind: string }
  | { type: 'stream_delta'; index: number; kind: 'text' | 'thinking'; text: string }
  | { type: 'stream_block_stop'; index: number }
  | { type: 'compact_boundary'; trigger: string | null; pre_tokens: number | null; post_tokens: number | null; duration_ms: number | null; timestamp?: number }
  | { type: 'error'; message: string }
  | { type: 'pong' };

export type ContentBlock =
  | { type: 'text'; text: string }
  | { type: 'thinking'; text: string }
  | { type: 'tool_use'; id: string; name: string; input: Record<string, unknown> }
  | { type: 'tool_result'; tool_use_id: string; content: ToolResultContent; is_error: boolean };

export type ToolResultContent =
  | string
  | Array<{ type: 'text'; text: string } | { type: string; [k: string]: unknown }>
  | null;

// ============== Chat REST: POST /chat/answer ==============

/** POST /chat/answer 请求 body */
export interface AnswerQuestionRequest {
  uuid: string;
  tool_use_id: string;
  answers: Record<string, string>;
  annotations?: Record<string, { preview?: string; notes?: string }>;
}

// ============== Shell WebSocket: /ws/shell ==============

export type ShellClientMessage =
  | { type: 'init'; cwd: string; shell?: string; cols: number; rows: number; token?: string }
  | { type: 'input'; data: string }
  | { type: 'resize'; cols: number; rows: number }
  | { type: 'signal'; signal: 'SIGINT' | 'SIGTERM' | 'SIGKILL' };

export type ShellServerMessage =
  | { type: 'ready' }
  | { type: 'output'; data: string }
  | { type: 'exit'; code: number }
  | { type: 'error'; message: string }
  | { type: 'cwd'; cwd: string };
